### ISSUE TITLE
Nested task spawning broken

### ISSUE THREAD

**Issue by rcohendw** [Amanieu/asyncplusplus#37]
When I use nested tasks with async::spawn() and async::get(), the synchronization doesn't work properly. This happens pretty consistently, and I am forced to never spawn tasks from tasks, which can be very limiting.

In the example below, some of the get()s will return before the work has completed, causing bugs.

Aside from this, the library is very convenient and works well.

Thanks!
Rob

```
void Function1()
{
	async::task<void> tasks[100];
	for (int t=0; t<100; t++)
	{
		tasks[t] = async::spawn([=]
		{
			Function2();
		});
	}

	for (int t=0; t<100; t++)
		tasks[t].get();
}

void Function2()
{
	async::task<void> tasks[100];
	for (int t=0; t<100; t++)
	{
		tasks[t] = async::spawn([=]
		{
			DoWork();
		});
	}

	for (int t=0; t<100; t++)
		tasks[t].get();
}

```

**Comment by rcohendw** [comment]
Commenting out the task stealing in threadpool_scheduler.cpp avoids the bug...

```
/*
			// Try to steal a task
			if (task_run_handle t = steal_task(impl, thread_id)) {
				t.run();
				break;
			}
*/
```

Perhaps the task stealing breaks assumptions related to task synchronization?

As you probably know, task stealing can be very desirable to maintain good CPU utilization and, in some cases, to avoid starvation deadlocks on shared resources.

Thanks!
Rob

**Comment by Amanieu** [comment]
Can you provide a full .cpp file that you use for testing this? I will look into it.

**Comment by rcohendw** [comment]
This explains why the bug appears only for nested get()s.  That's when task stealing comes into play.  Again, the symptom is that get()s sometimes return before the task has completed.

**Comment by rcohendw** [comment]
> Can you provide a full .cpp file that you use for testing this? I will look into it.

It's part of a pretty big project.  Perhaps I can put a simple repro test together that I can share.

**Comment by rcohendw** [comment]
I'll go ahead and do that now.  Thanks for your help!

**Comment by rcohendw** [comment]
Here's a clip that reproduces the bug on most runs, but not every run.  It will crash saying either the mutex is used after it has been destroyed or memory was accessed after being deallocated.  Basically, code not being run in the proper order, due to get() not blocking until the task is complete.

```
#include "async++.h"

const int kTestCount = 10;
int main(int argc, char **argv)
{
	for (int x=0; x<100; x++)
	{
		async::task<void> itasks[kTestCount];
		for (int i=0; i<kTestCount; i++)
		{
			int i2 = i;

			itasks[i] = async::spawn(
			[=]
			{
				vector<int> jvalues(kTestCount, 0);
				async::task<void> jtasks[kTestCount];
				mutex mutex;

				for (int j=0; j<kTestCount; j++)
				{
					int j2 = j;

					jtasks[j] = async::spawn(
					[=, &jvalues, &mutex]
					{
						int repeat = rand()%50;
						for (int r=0; r<repeat; r++)
						{
							mutex.lock();
							for (int k=0; k<kTestCount; k++)
								jvalues[k]++;
							mutex.unlock();
						}
					});
				}

				for (int j=0; j<kTestCount; j++)
					jtasks[j].get();
			});
		}

		for (int j=0; j<kTestCount; j++)
			itasks[j].get();
	}

	return 0;
}
```

Thanks!

**Comment by rcohendw** [comment]
Here's a better example.

It's reliable with NESTED_ASYNC 0, and crashes with NESTED_ASYNC 1

```
#define NESTED_ASYNC 1
const int kTestCount = 10;
int main(int argc, char **argv)
{
	for (int x=0; x<100; x++)
	{
#if NESTED_ASYNC
		async::task<void> itasks[kTestCount];
#endif
		for (int i=0; i<kTestCount; i++)
		{
			int i2 = i;

#if NESTED_ASYNC
			itasks[i] = async::spawn(
			[=]
			{
#endif
				async::task<void> jtasks[kTestCount];
				mutex mutex;

				for (int j=0; j<kTestCount; j++)
				{
					int j2 = j;

					jtasks[j] = async::spawn(
					[=, &mutex]
					{
						int repeat = rand()%50;
						for (int r=0; r<repeat; r++)
						{
							mutex.lock();
							mutex.unlock();
						}
					});
				}

				for (int j=0; j<kTestCount; j++)
					jtasks[j].get();
#if NESTED_ASYNC
			});
#endif
		}

#if NESTED_ASYNC
		for (int j=0; j<kTestCount; j++)
			itasks[j].get();
#endif
	}

	return 0;
}

```

**Comment by rcohendw** [comment]
Did this clip reproduce the problem for you?  I am using MSVC2017/Windows.

**Comment by Amanieu** [comment]
Yes, I managed to reproduce the issue on Linux. I've been a bit busy lately, but I am looking into it.

**Comment by rcohendw** [comment]
> Yes, I managed to reproduce the issue on Linux. I've been a bit busy lately, but I am looking into it.

Great! Thanks for the update!

**Comment by rcohendw** [comment]
I just tried the fix.  It's more reliable now, but my code above still crashes in release mode on MSVC2017 if I increase kTestCount to 100.  Thanks!

**Comment by rcohendw** [comment]
Actually, a Rebuild All seemed to fix it! I'll do more testing...

**Comment by rcohendw** [comment]
Unfortunately, it looks like it still isn't 100% reliable.  I get a rare null reference exception in task_wait_event.h on notify_one():
```

	// Signal an event and wake up a sleeping thread
	void signal(int event)
	{
		std::unique_lock<std::mutex> lock(mutex());
		event_mask |= event;
		lock.unlock();
		cond().notify_one();
	}
```

**Comment by Amanieu** [comment]
I just pushed a fix, can you try again?

**Comment by rcohendw** [comment]
Thanks! On it.

**Comment by rcohendw** [comment]
So far so good! I think you got it!

**Comment by rcohendw** [comment]
If you have a moment, on the topic of nested tasks, I have a question.

In the past, with this pattern, I've run into a problem with less than optimal CPU utilization:


```
Go()
{
	for (int i=0; i<4; i++)
		toptasks[i] = DoSomeWork();

	WaitForTasks(toptasks);
}

DoSomeWork()
{
	work = ProduceWork();

	for (int j=0; j<100; j++)
		subtasks[j] = ConsumeWork(work)

	WaitForTasks(subtasks);
}
```

The problem is if I start consuming the work right away, it will starve work production (and give me poor CPU utilization).  But if I don't start consuming work right away, I won't get good CPU utilization during work production, since the initial work creation parallelism is low.

In the past, I have been able to avoid this problem by prioritizing producer tasks over consumer tasks, which allows me to schedule all work as fast as possible without starving work production. Do you have a recommendation for a good way to approach this with async++, since there are no priorities?

Thanks!

**Comment by Amanieu** [comment]
I don't really have any good advice for you here. I would think that consuming the work right away would be best, since any thread which runs out of work to do will steal another (possibly producer) task from another thread, which should avoid the poor CPU utilization problem.

**Comment by rcohendw** [comment]
> I don't really have any good advice for you here. I would think that consuming the work right away would be best, since any thread which runs out of work to do will steal another (possibly producer) task from another thread, which should avoid the poor CPU utilization problem.

I've run into CPU utilization problems before with other libraries without priorities.  With priorities, I've been able to specify the order that tasks should be scheduled.  That's actually really nice, because work will always be scheduled immediately and in the most optimal order that I tune for my needs (highest CPU utilization, lowest latency, lowest memory usage, etc.)

In practice, this has magically worked out pretty decently for me with .Net's TPL (even without task priorities), since it will quickly & automatically add worker threads when it sees low CPU utilization.

It has been trickier with async++.  It appears that task stealing plucks tasks from the opposite side of the queue.  Maybe that is causing less optimal behavior for my use case?  Speaking to your point, it seems like my producers are getting somewhat starved, resulting in low CPU utilization.

I usually see this type of thing on libraries that prioritize minimizing synchronization overhead over flexible scheduling.  In practice, it seems like the flexible scheduling is more of a problem for me.  Although I really like the convenience of using the TPL-style interface, now that I'm spoiled by that.  :)

Maybe I need to work on an async++ custom scheduler that supports priorities.  I was hoping that there's a design pattern out there that solves this without writing a custom scheduler.

More specifically, I think the ideal priority order would be something like...

1: Block 1 producer job
1: Block 1 producer job
1: Block 1 producer job
2: Block 1 consumer job
2: Block 1 consumer job
2: Block 1 consumer job
3: Block 2 producer job
3: Block 2 producer job
3: Block 2 producer job
4: Block 2 consumer job
4: Block 2 consumer job
4: Block 2 consumer job

With manually specified priorities, I could just lay it out that way and they would always complete with the ideal order for max CPU utilization, min latency, and min memory usage.

Then again, maybe there's something clever I can do with the existing interface?

**Comment by rcohendw** [comment]
Maybe group 2 consumers waiting on group 1 producers would do the trick.  I think I'll try that.

**Comment by rcohendw** [comment]
I ended up needing to change the threadpool scheduler to a global priority queue and ditching the local thread queues.  I think the local thread queues were contributing to my problems, since they randomized the queue order.  For my use case, the flexibility of priorities was a bigger deal than lockless queues.

I did run into another issue.  I still allowed threads to grab new work from the global queue while waiting.  When there are a lot of work producer jobs that would wait, that could cause occasional degenerate cases where the stack blows up.  Sometimes, a thread will produce, wait, produce, wait, produce wait, etc. resulting in more stack depth for every subsequent job.  Currently, I am avoiding this problem by laying out priorities to insure that work gets consumed earlier.  Produce, wait, consume, return.  Seems to work.

Unfortunately, passing priority parameters to spawn() required fairly pervasive changes...not just a plug-in scheduler change.

**Comment by Amanieu** [comment]
Could you maybe use "proxy" allocators which have an attached priority but redirect to the general priority queue allocator?

**Comment by Amanieu** [comment]
Regarding stack blowup you could keep a thread-local counter for the number of nested waits and stop stealing if the counter exceeds a threshold. This should solve the stack overflows but may reduce parallelism.

**Comment by rcohendw** [comment]
The trouble with thread local priority queues is that respecting the overall priorities would require checking the global queue and all the thread queues first.  It seems like that would be less efficient, rather than more efficient.

Yes, I think a stack depth counter would be a good solution to make it robust against stack overflows.  Better to sacrifice parallelism in those cases than to crash.  I'll keep that in mind if it turns out to be a problem.  In practice, I found that certain priority patterns blow the stack (all producers first) while other priority patterns are fine (produce, consume, produce consume.)  That limitation works for me.

**Comment by rcohendw** [comment]
I guess another way to say it is that it's nice if the produce,consume,produce,consume order is a soft constraint, so the scheduler can get started on new produce tasks as soon as threads become available.

Given my use case and the heap overhead of lambda functions, I'm not really sweating lockless synchronization.

I'm not aware of a better way to do that than using priorities, but I'm not aware of a lot of things. :)

Thanks again for your help and sharing your library!  The lambda approach saves a ton of work, and I really appreciate how the library is focused, lightweight, portable, and self-contained.  :)

**Comment by rcohendw** [comment]
FYI, there was a remaining starvation issue.

The main (root) thread was getting slowed down by worker threads.  This was reducing CPU utilization, since the main thread is an important producer.  I was able to fix it by dropping the worker thread OS priority a notch, although that relies on non-portable code.

I've also seen this problem addressed without priorities by reserving 1 worker thread, except when a non-worker thread is blocking on a wait.  The priority approach is simpler, though, and works well.  It also has the benefit of playing nicer with other processes.  Just not portable.  :\
